File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word.downcase
    @letters = canonicalize(word)
  end

  def match(list)
    list.select do |word|
      canonicalize(word) == @letters && @word != word.downcase
    end
  end

  def canonicalize(word)
    word.downcase.chars.sort
  end
end

File: javascript/anagram/anagram.js
function toSortedString(word) {
  return word.toLowerCase().split('').sort().toString();
}

function Anagram(actual_word) {
  var actual_sorted = toSortedString(actual_word);

  this.match = function match(test_words) {
    return test_words.filter(function(test_word) {
      test_sorted = toSortedString(test_word);

      return actual_sorted == test_sorted && actual_word != test_word;
    });
  };
}

module.exports = Anagram;
